codalab/lib/bundle_store.py

```diff
@@ -383,3 +395,138 @@ def cleanup(self, uuid, dry_run):
+                    if ends_with_ext:
+                        to_delete += [path]
+                        continue
+                    elif '.' in path:
+                        print >> sys.stderr, 'WARNING: File %s is likely junk.' % path
+            return to_delete
+
+
+        partitions, _ = path_util.ls(self.partitions)
+        trash_count = 0
+
+        for partition in partitions:
+            partition_path = os.path.join(self.partitions, partition, MultiDiskBundleStore.DATA_SUBDIRECTORY)
+            entries = map(lambda f: os.path.join(partition_path, f),
+                          reduce(lambda d,f: d + f, path_util.ls(partition_path)))
+            bundle_paths = filter(_is_bundle, entries)
+            other_paths = set(entries) - set(bundle_paths)
+
+            uuids = map(_get_uuid, bundle_paths)
+            db_bundles = model.batch_get_bundles(uuid=uuids)
+            db_bundle_by_uuid = dict()
+            for bundle in db_bundles:
+                db_bundle_by_uuid[bundle.uuid] = bundle
+
+            # Check both bundles and non-bundles and remove each
+            for to_delete in _check_bundle_paths(bundle_paths, db_bundle_by_uuid):
+                trash_count += 1
+                _delete_path(to_delete)
+            for to_delete in _check_other_paths(other_paths, db_bundle_by_uuid):
+                trash_count += 1
+                _delete_path(to_delete)
+
+
+            # Check for each bundle if we need to compute its data_hash
+            data_hash_recomputed = 0
+
+            for bundle_path in bundle_paths:
+                uuid = _get_uuid(bundle_path)
+                bundle = db_bundle_by_uuid.get(uuid, None)
+                if bundle == None:
+                    continue
+                if compute_data_hash or bundle.data_hash == None:
+                    dirs_and_files = path_util.recursive_ls(bundle_path) if os.path.isdir(bundle_path) else ([], [bundle_path])
+                    data_hash = '0x%s' % path_util.hash_directory(bundle_path, dirs_and_files)
+                    if bundle.data_hash == None:
+                        data_hash_recomputed += 1
+                        print >> sys.stderr, 'Giving bundle %s data_hash %s' % (bundle_path, data_hash)
+                        if force:
+                            db_update = dict(data_hash=data_hash)
+                            model.update_bundle(bundle, db_update)
+                    elif compute_data_hash and data_hash != bundle.data_hash:
+                        data_hash_recomputed += 1
+                        print >> sys.stderr, 'Bundle %s should have data_hash %s, actual digest is %s' % (bundle_path, bundle.data_hash, data_hash)
+                        if repair_hashes and force:
+                            db_update = dict(data_hash=data_hash)
+                            model.update_bundle(bundle, db_update)
+
+
+        print >> sys.stderr, 'Deleted %d objects from the bundle store' % trash_count
```

Comment:
I felt like having these summary statistics might be nice as well for the dry-run, but I can also hide them